File: p39.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<vector<int>> solution;
    vector<int> candidates;
    void generate(vector<int>& curr, int target, int start) {
        if(target == 0) {
            solution.push_back(curr);
        }
        for(int i = start; i < candidates.size() && candidates[i] <= target; ++i) {
            curr.push_back(candidates[i]);
            generate(curr, target - candidates[i], i);
            curr.pop_back();
        }
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        this->candidates = candidates;
        vector<int> tmp;
        generate(tmp, target, 0);
        return solution;
    }
};

File: p39.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()

        ans = []
        curr = []

        def dfs(n, start):
            print(n, start)
            if n == 0:
                ans.append(curr[:])
                return
            if n < 0:
                return
            for i in range(start, len(candidates)):
                curr.append(candidates[i])
                dfs(n-candidates[i], i)
                curr.pop()

        dfs(target, 0)

        return ans
